File: Desktop Viewer (Swift)/NSLoggerSW/NSLoggerSW/LoggerConnection.m
/*
 * LoggerConnection.m
 *
 * BSD license follows (http://www.opensource.org/licenses/bsd-license.php)
 * 
 * Copyright (c) 2010-2011 Florent Pillet <fpillet@gmail.com> All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * Redistributions of  source code  must retain  the above  copyright notice,
 * this list of  conditions and the following  disclaimer. Redistributions in
 * binary  form must  reproduce  the  above copyright  notice,  this list  of
 * conditions and the following disclaimer  in the documentation and/or other
 * materials  provided with  the distribution.  Neither the  name of  Florent
 * Pillet nor the names of its contributors may be used to endorse or promote
 * products  derived  from  this  software  without  specific  prior  written
 * permission.  THIS  SOFTWARE  IS  PROVIDED BY  THE  COPYRIGHT  HOLDERS  AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
 * NOT LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A  PARTICULAR PURPOSE  ARE DISCLAIMED.  IN  NO EVENT  SHALL THE  COPYRIGHT
 * HOLDER OR  CONTRIBUTORS BE  LIABLE FOR  ANY DIRECT,  INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY,  OR CONSEQUENTIAL DAMAGES (INCLUDING,  BUT NOT LIMITED
 * TO, PROCUREMENT  OF SUBSTITUTE GOODS  OR SERVICES;  LOSS OF USE,  DATA, OR
 * PROFITS; OR  BUSINESS INTERRUPTION)  HOWEVER CAUSED AND  ON ANY  THEORY OF
 * LIABILITY,  WHETHER  IN CONTRACT,  STRICT  LIABILITY,  OR TORT  (INCLUDING
 * NEGLIGENCE  OR OTHERWISE)  ARISING  IN ANY  WAY  OUT OF  THE  USE OF  THIS
 * SOFTWARE,   EVEN  IF   ADVISED  OF   THE  POSSIBILITY   OF  SUCH   DAMAGE.
 * 
 */
#include <netinet/in.h>
#import <objc/runtime.h>
#import "LoggerConnection.h"
#import "LoggerMessage.h"
#import "LoggerCommon.h"
#import "MessageListener-Swift.h"

char sConnectionAssociatedObjectKey = 1;

@implementation LoggerConnection

@synthesize connectionInfo = _connectionInfo;
@synthesize delegate = _delegate;

- (id)init
{
	if ((self = [super init]) != nil)
	{
		_messages = [[NSMutableArray alloc] initWithCapacity:1024];
		self.parentIndexesStack = [[NSMutableArray alloc] init];
		_filenames = [[NSMutableSet alloc] init];
		_functionNames = [[NSMutableSet alloc] init];
	}
	return self;
}

- (id)initWithAddress:(NSData *)anAddress
{
	if ((self = [super init]) != nil)
	{
		_messages = [[NSMutableArray alloc] initWithCapacity:1024];
		self.parentIndexesStack = [[NSMutableArray alloc] init];
		_clientAddress = [anAddress copy];
		_filenames = [[NSMutableSet alloc] init];
		_functionNames = [[NSMutableSet alloc] init];
	}
	return self;
}

- (void)dealloc
{
}

- (BOOL)isNewRunOfClient:(LoggerConnection *)aConnection
{
	// Try to detect if a connection is a new run of an older, disconnected session
	// (goal is to detect restarts, so as to replace logs in the same window)
	assert(self.restoredFromSave == NO);

	// exclude files loaded from disk
	if (aConnection.restoredFromSave)
		return NO;
	
	// as well as still-up connections
	if (aConnection.connected)
		return NO;

	// check whether client info is the same
	BOOL (^isSame)(NSString *, NSString *) = ^(NSString *s1, NSString *s2)
	{
		if ((s1 == nil) != (s2 == nil))
			return NO;
		if (s1 != nil && ![s2 isEqualToString:s1])
			return NO;
		return YES;	// s1 and d2 either nil or same
	};

	if (!isSame(self.clientName, aConnection.clientName) ||
		!isSame(self.clientVersion, aConnection.clientVersion) ||
		!isSame(self.clientOSName, aConnection.clientOSName) ||
		!isSame(self.clientOSVersion, aConnection.clientOSVersion) ||
		!isSame(self.clientDevice, aConnection.clientDevice))
	{
		return NO;
	}
	
	// check whether address is the same, OR hardware ID (if present) is the same.
	// hardware ID wins (on desktop, iOS simulator can connect have different
	// addresses from run to run if the computer has multiple network interfaces / VMs installed
	if (self.clientUDID != nil && isSame(self.clientUDID, aConnection.clientUDID))
		return YES;
	
	if ((self.clientAddress != nil) != (aConnection.clientAddress != nil))
		return NO;

	if (self.clientAddress != nil)
	{
		// compare address blocks sizes (ipv4 vs. ipv6)
		NSUInteger addrSize = [self.clientAddress length];
		if (addrSize != [aConnection.clientAddress length])
			return NO;
		
		// compare ipv4 or ipv6 address. We don't want to compare the source port,
		// because it will change with each connection
		if (addrSize == sizeof(struct sockaddr_in))
		{
			struct sockaddr_in addra, addrb;
			[self.clientAddress getBytes:&addra length:sizeof(struct sockaddr_in)];
			[aConnection.clientAddress getBytes:&addrb length:sizeof(struct sockaddr_in)];
			if (memcmp(&addra.sin_addr, &addrb.sin_addr, sizeof(addra.sin_addr)))
				return NO;
		}
		else if (addrSize == sizeof(struct sockaddr_in6))
		{
			struct sockaddr_in6 addr6a, addr6b;
			[self.clientAddress getBytes:&addr6a length:sizeof(struct sockaddr_in6)];
			[aConnection.clientAddress getBytes:&addr6b length:sizeof(struct sockaddr_in6)];
			if (memcmp(&addr6a.sin6_addr, &addr6b.sin6_addr, sizeof(addr6a.sin6_addr)))
				return NO;
		}
		else if (![self.clientAddress isEqualToData:aConnection.clientAddress])
			return NO;		// we only support ipv4 and ipv6, so this should not happen
	}
	
	return YES;
}

- (void)messagesReceived:(NSArray *)msgs
{
    if (self.delegate) {
        [self.delegate connection:self didReceiveMessages:msgs];
    } else {
        NSLog(@"connection %@ has no delegate", self);
    }

}

- (void)clearMessages
{
    // Clear the backlog of messages, only keeping the top (client info) message
    // This MUST be called on the messageProcessingQueue
    if (![self.messages count])
        return;

    // Locate the clientInfo message
    if (((LoggerMessage *)[self.messages objectAtIndex:0]).type == LOGMSG_TYPE_CLIENTINFO)
        [self.messages removeObjectsInRange:NSMakeRange(1, [self.messages count]-1)];
    else
        [self.messages removeAllObjects];
}

- (void)clientInfoReceived:(LoggerMessage *)message
{
	// Insert message at first position in the message list. In the unlikely event there is
	// an existing ClientInfo message at this position, just replace it. Also, don't fire
	// a "didReceiveMessages". The rationale behind this is that if the connection just came in,
	// we are not yet attached to a window and when attaching, the window will refresh all messages.
    if ([self.messages count] == 0 || ((LoggerMessage *)[self.messages objectAtIndex:0]).type != LOGMSG_TYPE_CLIENTINFO)
        [self.messages insertObject:message atIndex:0];

	// all this stuff occurs on the main thread to avoid touching values
	// while the UI reads them
    NSDictionary *parts = message.parts;
    id value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_CLIENT_NAME]];
    if (value != nil)
        self.clientName = value;
    value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_CLIENT_VERSION]];
    if (value != nil)
        self.clientVersion = value;
    value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_OS_NAME]];
    if (value != nil)
        self.clientOSName = value;
    value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_OS_VERSION]];
    if (value != nil)
        self.clientOSVersion = value;
    value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_CLIENT_MODEL]];
    if (value != nil)
        self.clientDevice = value;
    value = [parts objectForKey:[NSNumber numberWithInteger:PART_KEY_UNIQUEID]];
    if (value != nil)
        self.clientUDID = value;

}

- (NSString *)clientAppDescription
{
	// enforce thread safety (only on main thread)
	assert([NSThread isMainThread]);
	NSMutableString *s = [[NSMutableString alloc] init];
	if (self.clientName != nil)
		[s appendString:self.clientName];
	if (self.clientVersion != nil)
		[s appendFormat:@" %@", self.clientVersion];
	if (self.clientName == nil && self.clientVersion == nil)
		[s appendString:NSLocalizedString(@"<unknown>", @"")];
	if (self.clientOSName != nil && self.clientOSVersion != nil)
		[s appendFormat:@"%@(%@ %@)", [s length] ? @" " : @"", self.clientOSName, self.clientOSVersion];
	else if (self.clientOSName != nil)
		[s appendFormat:@"%@(%@)", [s length] ? @" " : @"", self.clientOSName];

	return s;
}

- (NSString *)clientAddressDescription
{
	// subclasses should implement this
	return @"";
}

- (NSString *)clientDescription
{
	// enforce thread safety (only on main thread)
	assert([NSThread isMainThread]);
	return [NSString stringWithFormat:@"%@ @ %@", [self clientAppDescription], [self clientAddressDescription]];
}

- (NSString *)status
{
	// status is being observed by LoggerStatusWindowController and changes once
	// when the connection gets disconnected
	NSString *format;
	if (self.connected)
		format = NSLocalizedString(@"%@ connected", @"");
	else
		format = NSLocalizedString(@"%@ disconnected", @"");
	if ([NSThread isMainThread])
		return [NSString stringWithFormat:format, [self clientDescription]];
	__block NSString *status;
	dispatch_sync(dispatch_get_main_queue(), ^{
		status = [NSString stringWithFormat:format, [self clientDescription]];
	});
	return status;
}

- (void)setConnected:(BOOL)newConnected
{
	if (_connected != newConnected)
	{
		_connected = newConnected;
		
		if (!self.connected && [(id)self.delegate respondsToSelector:@selector(remoteDisconnected:)])
			[(id)self.delegate performSelectorOnMainThread:@selector(remoteDisconnected:) withObject:self waitUntilDone:NO];

//		[[NSNotificationCenter defaultCenter] postNotificationName:kShowStatusInStatusWindowNotification
//															object:self];
	}
}

- (void)shutdown
{
	self.connected = NO;
}

- (NSDictionary*)connectionInfo
{
    if (! _connectionInfo) {

        NSMutableDictionary* tmpConnectionInfo = [NSMutableDictionary dictionary];
        _connectionInfo = tmpConnectionInfo;

        tmpConnectionInfo[@"clientName"]      = self.clientName;
        tmpConnectionInfo[@"clientVersion"]   = self.clientVersion;
        tmpConnectionInfo[@"clientOSName"]    = self.clientOSName;
        tmpConnectionInfo[@"clientOSVersion"] = self.clientOSVersion;
        tmpConnectionInfo[@"clientDevice"]    = self.clientDevice;
        tmpConnectionInfo[@"clientUDID"]      = self.clientUDID;
    }

    return _connectionInfo;
}

// -----------------------------------------------------------------------------
#pragma mark -
#pragma mark NSCoding
// -----------------------------------------------------------------------------
- (id)initWithCoder:(NSCoder *)aDecoder
{
	if ((self = [super init]) != nil)
	{
		self.clientName = [aDecoder decodeObjectForKey:@"clientName"];
		self.clientVersion = [aDecoder decodeObjectForKey:@"clientVersion"];
		self.clientOSName = [aDecoder decodeObjectForKey:@"clientOSName"];
		self.clientOSVersion = [aDecoder decodeObjectForKey:@"clientOSVersion"];
		self.clientDevice = [aDecoder decodeObjectForKey:@"clientDevice"];
		self.clientUDID = [aDecoder decodeObjectForKey:@"clientUDID"];
		self.parentIndexesStack = [aDecoder decodeObjectForKey:@"parentIndexes"];
		_filenames = [aDecoder decodeObjectForKey:@"filenames"];
		if (_filenames == nil)
			_filenames = [[NSMutableSet alloc] init];
		_functionNames = [aDecoder decodeObjectForKey:@"functionNames"];
		if (_functionNames == nil)
			_functionNames = [[NSMutableSet alloc] init];
		objc_setAssociatedObject(aDecoder, &sConnectionAssociatedObjectKey, self, OBJC_ASSOCIATION_ASSIGN);
		_messages = [aDecoder decodeObjectForKey:@"messages"];
		_reconnectionCount = [aDecoder decodeIntForKey:@"reconnectionCount"];
		_restoredFromSave = YES;
		
	}
	return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder
{
	if (self.clientName != nil)
		[aCoder encodeObject:self.clientName forKey:@"clientName"];
	if (self.clientVersion != nil)
		[aCoder encodeObject:self.clientVersion forKey:@"clientVersion"];
	if (self.clientOSName != nil)
		[aCoder encodeObject:self.clientOSName forKey:@"clientOSName"];
	if (self.clientOSVersion != nil)
		[aCoder encodeObject:self.clientOSVersion forKey:@"clientOSVersion"];
	if (self.clientDevice != nil)
		[aCoder encodeObject:self.clientDevice forKey:@"clientDevice"];
	if (self.clientUDID != nil)
		[aCoder encodeObject:self.clientUDID forKey:@"clientUDID"];
	[aCoder encodeObject:self.filenames forKey:@"filenames"];
	[aCoder encodeObject:self.functionNames forKey:@"functionNames"];
	[aCoder encodeInt:self.reconnectionCount forKey:@"reconnectionCount"];
	@synchronized (self.messages)
	{
		[aCoder encodeObject:self.messages forKey:@"messages"];
		[aCoder encodeObject:self.parentIndexesStack forKey:@"parentIndexes"];
	}
}

- (void)setDelegate:(id<LoggerConnectionDelegate>)delegate
{
    _delegate = delegate;
}

- (id<LoggerConnectionDelegate>)delegate {
    return _delegate;
}

@end

File: Desktop Viewer (Swift)/NSLoggerSW/NSLoggerSW/LoggerConnection.swift
//
//  LoggerConnection.swift
//  NSLoggerSW
//
//  Created by Guillaume Laurent on 21/04/15.
//  Copyright (c) 2015 Guillaume Laurent. All rights reserved.
//

import Cocoa
import Dispatch

var sConnectionAssociatedObjectKey:UInt8 = 1

let kShowStatusInStatusWindowNotification = "ShowStatusInStatusWindowNotification"

protocol LoggerConnectionDelegate {

    func connection(connection:LoggerConnection, messages didReceiveMessages:[LoggerMessage], range rangeInMessagesList:Range<Int>)
    func remoteDisconnected(connection:LoggerConnection)

}

func ==(a:sockaddr_in, b:sockaddr_in) -> Bool {

    var a1 = a
    var b1 = b

    return memcmp(&a1, &b1, sizeof(sockaddr_in)) == 0
}

func ==(a:sockaddr_in6, b:sockaddr_in6) -> Bool {

    var a1 = a
    var b1 = b

    return memcmp(&a1, &b1, sizeof(sockaddr_in6)) == 0
}


class LoggerConnection: NSObject, NSCoding {

    var clientName:String?
    var clientVersion:String?
    var clientOSName:String?
    var clientOSVersion:String?
    var clientDevice:String?
    var clientUDID:String?

    var filenames = Set<String>()
    var functionNames = Set<String>()

    var clientAddress:NSData?

    var messages = [LoggerMessage]()
    var parentIndexesStack = [Int]()

    var messageProcessingQueue = dispatch_queue_create("com.florentpillet.nslogger.messageProcessingQueue", nil)

    var reconnectionCount = 0

    var connected:Bool {
        didSet {
            if !connected {
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    self.delegate?.remoteDisconnected(self)
                })
            }
        }
    }

    var restoredFromSave = false
    var attachedToWindow = false

    var delegate:LoggerConnectionDelegate?

    required override init() {
        connected = false
        super.init()
    }

    required init(coder aDecoder: NSCoder) {

        connected = false

        super.init()

        clientName         = aDecoder.decodeObjectForKey("clientName") as? String
        clientVersion      = aDecoder.decodeObjectForKey("clientVersion") as? String
        clientOSName       = aDecoder.decodeObjectForKey("clientOSName") as? String
        clientOSVersion    = aDecoder.decodeObjectForKey("clientOSVersion") as? String
        clientDevice       = aDecoder.decodeObjectForKey("clientDevice") as? String
        clientUDID         = aDecoder.decodeObjectForKey("clientUDID") as? String
        parentIndexesStack = (aDecoder.decodeObjectForKey("parentIndexes") as? [Int])!

        if let decodedFileNames = aDecoder.decodeObjectForKey("filenames") as? Set<String> {
            filenames = decodedFileNames
        }

        if let decodedFunctionNames = aDecoder.decodeObjectForKey("functionNames") as? Set<String> {
            functionNames = decodedFunctionNames
        }

        objc_setAssociatedObject(aDecoder, &sConnectionAssociatedObjectKey, self, objc_AssociationPolicy(OBJC_ASSOCIATION_ASSIGN))

        if let decodedMessages = aDecoder.decodeObjectForKey("messages") as? [LoggerMessage] {
            messages = decodedMessages
        }

        reconnectionCount = aDecoder.decodeIntegerForKey("reconnectionCount")

        restoredFromSave = true

        messageProcessingQueue = dispatch_queue_create("com.florentpillet.nslogger.messageProcessingQueue", nil)

    }

    func encodeWithCoder(aCoder: NSCoder) {
        if let clientName = self.clientName {
            aCoder.encodeObject(clientName, forKey: "clientName")
        }
        if let clientVersion = self.clientVersion {
            aCoder.encodeObject(clientVersion, forKey: "clientVersion")
        }
        if let clientOSName = self.clientOSName {
            aCoder.encodeObject(clientName, forKey: "clientOSName")
        }
        if let clientOSVersion = self.clientOSVersion {
            aCoder.encodeObject(clientOSVersion, forKey: "clientOSVersion")
        }
        if let clientDevice = self.clientDevice {
            aCoder.encodeObject(clientDevice, forKey: "clientDevice")
        }
        if let clientUDID = self.clientUDID {
            aCoder.encodeObject(clientUDID, forKey: "clientUDID")
        }

        aCoder.encodeObject(filenames, forKey: "filenames")
        aCoder.encodeObject(functionNames, forKey: "functionNames")
        aCoder.encodeInteger(reconnectionCount, forKey: "reconnectionCount")

        synced(messages) { () -> () in
            aCoder.encodeObject(self.messages, forKey: "messages")
            aCoder.encodeObject(self.parentIndexesStack, forKey: "parentIndexes")
        }
    }


    func isNewRunOfClient(connection:LoggerConnection) -> Bool {

        assert(restoredFromSave == false, "isNewRunOfClient called with restoredFromSave == true")

        if connection.restoredFromSave {
            return false
        }

        if connection.connected {
            return false
        }

        if clientName != connection.clientName ||
            clientVersion != connection.clientVersion ||
            clientOSName != connection.clientOSName ||
            clientOSVersion != connection.clientOSVersion ||
            clientDevice != connection.clientDevice {
                return false
        }

        if clientUDID == connection.clientUDID {
            return true
        }

        if (clientAddress != nil) != (connection.clientAddress != nil) {
            return false
        }

        if let clientAddress = self.clientAddress, connectionClientAddress = connection.clientAddress {

            let addrSize = clientAddress.length

            if addrSize != connection.clientAddress?.length {
                return false
            }

            if addrSize == sizeof(sockaddr_in) {
                var addrA = sockaddr_in()
                var addrB = sockaddr_in()

                clientAddress.getBytes(&addrA, length:addrSize)
                connectionClientAddress.getBytes(&addrB, length:addrSize)

                if !(addrA == addrB) {
                    return false
                }

            } else if addrSize == sizeof(sockaddr_in6) {
                var addrA = sockaddr_in6()
                var addrB = sockaddr_in6()

                clientAddress.getBytes(&addrA, length:addrSize)
                connectionClientAddress.getBytes(&addrB, length:addrSize)

                if !(addrA == addrB) {
                    return false
                }

            } else {
                if !clientAddress.isEqualToData(connectionClientAddress) {
                    return false
                }
            }

        }

        return true
    }


    func messagesReceived(theMessages:[LoggerMessage]) {
        dispatch_async(messageProcessingQueue) { () -> Void in

            var range:Range<Int> = Range(start: self.messages.count, end: theMessages.count)

            synced(theMessages) { () -> () in
                self.messages += theMessages
            }

            if self.attachedToWindow {
                self.delegate?.connection(self, messages: theMessages, range: range)
            }

        }
    }

    func clearMessages() {
        dispatch_sync(messageProcessingQueue, { () -> Void in
            if self.messages.count > 0 {
                if self.messages[0].type == .ClientInfo {
                    let range = Range<Int>(start: 1, end: self.messages.count - 1)
                    self.messages.removeRange(range)
                } else {
                    self.messages.removeAll(keepCapacity: false)
                }
            }
        })
    }

    func clientInfoReceived(message:LoggerMessage) {
        dispatch_async(messageProcessingQueue) { () -> Void in
            synced(self.messages) { () -> () in
                if self.messages.count == 0 || self.messages[0].type != .ClientInfo {
                    self.messages.insert(message, atIndex: 0)
                }
            }
        }

        dispatch_async(dispatch_get_main_queue()) { () -> Void in
            let parts = message.parts

            if let value = parts[.ClientName] as? String {
                self.clientName = value
            }

            if let value = parts[.ClientVersion] as? String {
                self.clientVersion = value
            }

            if let value = parts[.OSName] as? String {
                self.clientOSName = value
            }

            if let value = parts[.OSVersion] as? String {
                self.clientOSVersion = value
            }

            if let value = parts[.ClientModel] as? String {
                self.clientDevice = value
            }

            if let value = parts[.UniqueId] as? String {
                self.clientUDID = value
            }

            NSNotificationCenter.defaultCenter().postNotificationName(kShowStatusInStatusWindowNotification, object: self)

        }

    }


    func clientAppDescription() -> String {
        assert(NSThread.isMainThread())

        var res = ""

        if let clientName = self.clientName {
            res += clientName
        }

        if let clientVersion = self.clientVersion {
            res += " \(clientVersion)"
        }

        if clientName == nil && clientVersion == nil {
            res = NSLocalizedString("<unknown>", comment:"")
        }

        if let clientOSName = self.clientOSName {

            if let clientOSVersion = self.clientOSVersion {
                if !res.isEmpty {
                    res += " "
                }
                res += "\(clientOSName) \(clientOSVersion)"
            } else {
                if !res.isEmpty {
                    res += " "
                }
                res += "\(clientOSName)"
            }
        }

        return res
    }

    func clientAddressDescription() -> String {
        return ""
    }

    func clientDescription() -> String {
        assert(NSThread.isMainThread())

        return "\(clientAppDescription()) @ \(clientAddressDescription())"
    }

    func status() -> String {
        let connectedString = connected ? NSLocalizedString("connected", comment: "") : NSLocalizedString("disconnected", comment: "")

        var res = ""

        dispatch_sync(dispatch_get_main_queue()) { () -> Void in
            res = "\(self.clientDescription()) \(connectedString)"
        }

        return res
    }


    func shutdown() {
        connected = false
    }

}
